### DISCUSSION TITLE
Some ideas to improve the project

### DISCUSSION THREAD

**Discussion by nekoxuee** [Samillion/ModernZ#209] [Ideas]
Hi,

I've thought of a few ideas that I think could help improve the project. Since this is a rewrite, I think it's a good opportunity to establish a solid structure that could serve as a foundation for future forks or contributors.

1. **Establishing a clear vision**

   As a ModernZ user, I often make changes but hesitate to submit a PR because I'm unsure if they align with the project goal. A clear project vision would be nice for contributors to know what to focus on.

2. **Implement conventions**

   Commit message guidelines: Adopting a format like [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) to make it easier to understand the history and purpose of changes.

   Branching strategy: Clearly defining how feature and fix branches should be named and managed can simplify collaboration.

3. **release drafter**

   I came across [release drafter](https://github.com/release-drafter/release-drafter) and think it could be handy for automating release notes based on commits. It’d keep users updated with clean changelogs.

Let me know what you think about these ideas.

**Comment by Samillion** [comment]
> As a ModernZ user, I often make changes but hesitate to submit a PR because I'm unsure if they align with the project goal. A clear project vision would be nice for contributors to know what to focus on.

Answering you specifically, never hesitate. I've already seen your work and have a glimpse of what you think is practical. Worst case scenario is that your PR will be a discussion, but your ideas and implementations have been solid.

To answer generally for all contributors, the goal is to establish a stable OSC alternative to stock with a modern look. Practical use is the key vision. Niche features or additions like https://github.com/Samillion/ModernZ/issues/84 can be considered, but they're definitely not the main priority. 

At least while the project is still fixing very old bugs from [previous forks or origin](https://github.com/Samillion/ModernZ/issues/87#issuecomment-2433502772).

**Main project goals:**
- Revive and maintain the Modern OSC, as it is either abandoned or partially maintained by adding new code without fixing the old.
- Apply common features that are missing and most users would want (ie: customization, button functions, lists and so on)

ModernZ is trying to be in a middle area between stock OSC and a project like uosc. One is too simple, the other is too complicated.

One of the first things that happened in ModernZ was to re-base on stock osc and integrate features that mpv already has, which eliminated a ton of bloat from the code. (example: https://github.com/Samillion/ModernZ/commit/cbce79c9cf378e19dc1417304e2d6fc0bc69f282)

In essence, we want to be better than stock OSC, but we don't want to do everything. That is why for example I loved that you integrated locale as extras, and the same thing was done for the pause indicator. If it's useful, but not necessary to the OSC, it should be in extras or an independent script. Modular methodology.

> Commit message guidelines: Adopting a format like [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) to make it easier to understand the history and purpose of changes.

I absolutely agree. I was hectic in the beginning, but recently I've been applying the same commit format as mpv. 

> Branching strategy: Clearly defining how feature and fix branches should be named and managed can simplify collaboration.

I honestly haven't thought of this much, what I've been doing is: if it's for testing, any random (yet relevant naming scheme) to branch would suffice, if it's a branch point that I know might be merged, I implement a clear naming related to the commit message with the prefix `dev_`.

Which I guess we can utilize as commit conventions, ie: `dev_`, `fix_`, `feat_`

> I came across [release drafter](https://github.com/release-drafter/release-drafter) and think it could be handy for automating release notes based on commits. It’d keep users updated with clean changelogs.

That is wonderful, and I agree, it would make things a lot smoother and seamless. Not just for users, but for maintainers and contributors as well.

Thank you so much for your detailed input. If you have any ideas of something that you'd think should be implemented, please share it and never hesitate.

**Comment by nekoxuee** [comment]
tbh, I’m not that confident in my lua skills yet. I mostly know the basics because of javascript, and I often find myself trying things until they work. That said, I’m always happy to help however I can, whether it’s with ideas, testing, docs, or anything else you might need for the project.

Here are some changes I’ve made to my version of ModernZ that you might find interesting:

- **Picture-in-Picture mode for "On Top"**: I modified the pin button to act as a picture-in-picture mode. It not only pins the window on top, but also resizes it to the bottom corner of the screen, which I find useful for multitasking.
  
- **Updated icon font**: Since the original ModernX was inspired by the Windows 10 Movies & TV app, I decided to switch to fluent icons, because it felt like a natural progression. Currently, ModernZ uses an older version of Material Icons. In 2022 Google introduced Material Symbols, which is the current maintained set.

- **Moved playlist button to title bar**: I relocated the playlist button to the window’s title bar, just to the left of the title. Many modern players follow this design, and I think it visually balances the interface by not clustering all buttons in the bottom row.

- **Back button for stopping playback**: I recently thought of adding a back button in the left corner of the title bar, next to the playlist button. This button stops playback and returns to idle mode.

If any of these features sound useful, I’d be glad to discuss them further or help implement them. And thanks for the detailed response

**Comment by Samillion** [comment]
> I’m always happy to help however I can, whether it’s with ideas, testing, docs, or anything else you might need for the project.

Much appreciated, your ideas and implementations have definitely been very helpful.

> Picture-in-Picture mode for "On Top": I modified the pin button to act as a picture-in-picture mode. It not only pins the window on top, but also resizes it to the bottom corner of the screen, which I find useful for multitasking.

Great idea, extremely useful for single screen use. We could implement it as a mode `pip, pin` for the ontop button, to control behavior, in order to satisfy both scenarios. (ie: pip in same screen, or pin in another screen or wherever)

> Updated icon font: Since the original ModernX was inspired by the Windows 10 Movies & TV app, I decided to switch to fluent icons, because it felt like a natural progression. Currently, ModernZ uses an older version of Material Icons. In 2022 Google introduced Material Symbols, which is the current maintained set.

My god, you have read my mind. That is wonderful! I cannot wait to see it. If you do a PR on this, please include a font credit to yourself in "History & Credits". In fact, you should include one for the locales as well. As both are major features within the osc.

Speaking of locales, after I cleanup bugs and code, I plan (try) on turning the locales into JSON file to be used within `script-opts`. For the sake of staying true to locale standards and to avoid loading them as a script automatically in mpv's backend.

> Moved playlist button to title bar: I relocated the playlist button to the window’s title bar, just to the left of the title. Many modern players follow this design, and I think it visually balances the interface by not clustering all buttons in the bottom row.

Indeed, I think I've seen it in MPV Easy and even some frontends like Celluloid. So this makes the playlist button visible in `--no-border` or `--top-bar=no` only?

> Back button for stopping playback: I recently thought of adding a back button in the left corner of the title bar, next to the playlist button. This button stops playback and returns to idle mode.

That sounds very interesting, it actually behaves like a panic button as well, which I'm sure many would find useful.

**Comment by nekoxuee** [comment]
> Great idea, extremely useful for single screen use. We could implement it as a mode pip, pin for the ontop button, to control behavior, in order to satisfy both scenarios. (ie: pip in same screen, or pin in another screen or wherever)

Just to clarify, it sounds like you’re suggesting keeping both the current "On Top" behaviour and adding the new picture-in-picture functionality as a user option, so they can choose between the two modes?

<br>

> My god, you have read my mind. That is wonderful! I cannot wait to see it. If you do a PR on this

In my initial comment, I mentioned both Fluent Icons and Material Symbols, and I’m not sure which you prefer for the final look.
To clarify, I didn’t create the .ttf file myself. I used Photoshop to edit it, by adding the missing jump icons for 5 and 30 seconds. I also included a few filled icon variants from Microsoft’s Figma library.

<br>

> Speaking of locales, after I cleanup bugs and code, I plan (try) on turning the locales into JSON file to be used within script-opts. For the sake of staying true to locale standards and to avoid loading them as a script automatically in mpv's backend.

Moving to JSON for locales is a great idea. After implementing the current approach, I realized JSON would be a better way to do it.

<br>

> Indeed, I think I've seen it in MPV Easy and even some frontends like Celluloid. So this makes the playlist button visible in --no-border or --top-bar=no only?

I think it might be best to keep the playlist button visible regardless of `--no-border` or `--top-bar=no`, so users can have the button without needing to disable the border.

**Comment by Samillion** [comment]
> Just to clarify, it sounds like you’re suggesting keeping both the current "On Top" behaviour and adding the new picture-in-picture functionality as a user option, so they can choose between the two modes?

Wouldn't that be the best scenario? For example, on the PC where I do most of my work, I have dual screens. If I watch something, I tend to to do it like this:

![image](https://github.com/user-attachments/assets/fe088667-1a4d-4b5a-b00c-ffbc2cad8d63)

Which gives me more flexibility when I pin in the second screen.

However, on my other PCs, I have a single screen, so a pip mode would be best in that scenario.

> In my initial comment, I mentioned both Fluent Icons and Material Symbols, and I’m not sure which you prefer for the final look.

I meant I was excited to see your work with the Fluent Icons. I was sneaky 😝 and looked over your mpv conf repo and noticed the work with the font and icons. Beautiful.

> I didn’t create the .ttf file myself. I used Photoshop to edit it, by adding the missing jump icons for 5 and 30 seconds. I also included a few filled icon variants from Microsoft’s Figma library.

Still deserve credit for the effort, to be honest. A font origin credit to be included would solve this, so that everyone's effort are acknowledged.

> Moving to JSON for locales is a great idea. After implementing the current approach, I realized JSON would be a better way to do it.

To be honest, your PR is what made me think about the possibility of JSON. Before your PR I didn't even think of integrating it further, as you saw. 😅 

> I think it might be best to keep the playlist button visible regardless of `--no-border` or `--top-bar=no`, so users can have the button without needing to disable the border.

That would be very interesting. I'd love to see it in action, as I might be confused on how this idea is implemented.

**Comment by ThousandDemonDaggers** [comment]
Will this project follows and implement features and fixes added to ModernX? For example the new 0.3.8 release.

**Comment by Samillion** [comment]
> Will this project follows and implement features and fixes added to ModernX? For example the new 0.3.8 release.

The dynamic time feature is by @Xurdejl which he implemented in a PR to both ModernZ and ModernX.

All ModernX (and Modern in general) features are already added. ModernZ is a project to revive the modern origin, so that means ModernZ will have less bugs and more customization than other forks. Even the download button was implemented recently, and will be released in a couple of days. (Can use it in this [version](https://github.com/Samillion/ModernZ/blob/main/modernz.lua))

The only features that will not be added are: Youtube description, comments and likes. Reason being those are niche features that should be used as independent scripts and not tied to the OSC, for maintenance sake. Similar to our Pause Indicator and Auto-ytdlFormat in [Extras](https://github.com/Samillion/ModernZ#extras).

**Comment by nekoxuee** [comment]
> Wouldn't that be the best scenario? For example, on the PC where I do most of my work, I have dual screens. If I watch something, I tend to to do it like this:

True, I didn't think about that because I only have one monitor.

<br>

> > I think it might be best to keep the playlist button visible regardless of --no-border or --top-bar=no, so users can have the button without needing to disable the border.
>
> That would be very interesting. I'd love to see it in action, as I might be confused on how this idea is implemented.

I meant to keep the icon in the top left corner, not necessarily within the title bar like in no-border.

**Comment by ThousandDemonDaggers** [comment]
> > Will this project follows and implement features and fixes added to ModernX? For example the new 0.3.8 release.
> 
> The dynamic time feature is by @Xurdejl which he implemented in a PR to both ModernZ and ModernX.
> 
> All ModernX (and Modern in general) features are already added. ModernZ is a project to revive the modern origin, so that means ModernZ will have less bugs and more customization than other forks. Even the download button was implemented recently, and will be released in a couple of days. (Can use it in this [version](https://github.com/Samillion/ModernZ/blob/main/modernz.lua))
> 
> The only features that will not be added are: Youtube description, comments and likes. Reason being those are niche features that should be used as independent scripts and not tied to the OSC, for maintenance sake. Similar to our Pause Indicator and Auto-ytdlFormat in [Extras](https://github.com/Samillion/ModernZ#extras).

Oh thanks for the clarifications! At this point both ModernZ and ModernX have become something amazing.

I really wonder if there is an idea of ​​merging the two projects into one and joining forces, having ModernZ as a base. Was this discussed with zydezu?

**Comment by Samillion** [comment]
Not in-depth, to be honest. This fork started by me trying to improve a specific issue on ModernX, then I discovered a few bugs and that it was based on old code.

I re-did the entire project to re-base it on stock OSC code (to ensure compatibility), and then it snowballed into massive changes within the code base to the point that ModernZ is now more compatible with stock OSC than any of the modern forks/origin, because they're still using the same old code base.

Reference: https://github.com/zydezu/ModernX/issues/49#issuecomment-2392634930

Having said that, all the other Modern forks and origin work beautifully. "old/base" difference is merely semantics. There might be bugs in them, but as far as I know it's not critical, mostly cosmetic or behavioral.

I agree there should be a combined effort, but it's honestly hard to coordinate.

One of the biggest wins of having different forks is that the user wins. You have many options to pick from, if you want something basic that works, go for modern origin. You want Youtube features (comments, description...etc), go with Zydezu's fork...etc

**Comment by Samillion** [comment]
> True, I didn't think about that because I only have one monitor.

You mean a pip behavior like this auto-profile, but for the pin button?
```ini
[Window-PIP]
    profile-desc=On window pin: resize, move window, enable persistant bar
    profile-cond=ontop and ontop == true and not fullscreen
    profile-restore=copy-equal
    script-opts-append=modernz-persistentprogress=yes
    geometry=100%:100%
    autofit=50%x40%
    autofit-larger=50%x40%
```

**Comment by nekoxuee** [comment]
> You mean a pip behavior like this auto-profile, but for the pin button?
> 
> ```ini
> [Window-PIP]
>     profile-desc=On window pin: resize, move window, enable persistant bar
>     profile-cond=ontop and ontop == true and not fullscreen
>     profile-restore=copy-equal
>     script-opts-append=modernz-persistentprogress=yes
>     geometry=100%:100%
>     autofit=50%x40%
>     autofit-larger=50%x40%
> ```

I've been busy the last days, but I had this: https://github.com/Xurdejl/mpv-config/commit/3983d710fa893d231efb9d895359f8e93fd6fc27. The osc scale needs to be changed because when resizing becomes too small.

**Comment by Samillion** [comment]
> I've been busy the last days, but I had this: [Xurdejl/mpv-config@3983d71](https://github.com/Xurdejl/mpv-config/commit/3983d710fa893d231efb9d895359f8e93fd6fc27). The osc scale needs to be changed because when resizing becomes too small.

Someone suggested a behavior for PIP that resembles Android PIP, what do you think? https://github.com/maoiscat/mpv-osc-modern/issues/38

Or, instead of changing the behavior for the minimize button, we can add next to minimize a "PIP" button, shows when window is pinned (no border)

**Comment by nekoxuee** [comment]
> Someone suggested a behavior for PIP that resembles Android PIP, what do you think? [maoiscat/mpv-osc-modern#38](https://github.com/maoiscat/mpv-osc-modern/issues/38)
> 
> Or, instead of changing the behavior for the minimize button, we can add next to minimize a "PIP" button, shows when window is pinned (no border)

Desktop video players usually support PIP as a separate button rather than tied to the minimize button, which aligns better with desktop UI norms. I feel linking it to the minimize button might be unintuitive for desktop users. So yeah, the separate button better.

For context, here are the UIs I took as an inspiration:

1. Windows Media Player
![Windows Media Player](https://github.com/user-attachments/assets/c76d43ca-bf4f-4855-993d-cbadfbe60be5)

2. Screenbox Media Player ([source](https://github.com/huynhsontung/Screenbox))
![Screenbox Media Player](https://github.com/user-attachments/assets/fa85da57-2c0a-4c40-9b63-2751e257c751)

**Comment by Keith94** [comment]
Any thoughts about moving this (and any future ideas/questions) into [discussions?](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/enabling-or-disabling-github-discussions-for-a-repository)

**Reply by Samillion** [reply]
Done!

**Comment by Samillion** [comment]
> Any thoughts about moving this (and any future ideas/questions) into [discussions?](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/enabling-or-disabling-github-discussions-for-a-repository)

Terribly sorry about the late response. The recent changes took all my attention.

I thought about discussions, I disabled it when I first started the repo because I didn't think it would be needed.

Should this repo enable Discussions or just rely on issues with labels?

> Screenbox Media Player ([source](https://github.com/huynhsontung/Screenbox))

I really like the seekbar and the volumebar on this, it's a bit thicker, and the border around the slider handle is kind of neat.

Should we mimic this on ModernZ?

Also, soon I'm going to add a new layout that would have the controls on the side similar to it as well. We then would have the option to use "center", "side" layouts.

**Comment by Keith94** [comment]
>Should this repo enable Discussions or just rely on issues with labels?

Just issues should be fine.. the amount of new ones is pretty small anyway.

> Should we mimic this on ModernZ?

My opinion yes, would solve #179 and the border is a nice touch.

**Comment by nekoxuee** [comment]
> Should we mimic this on ModernZ?

I think it would look neat.

> Also, soon I'm going to add a new layout that would have the controls on the side similar to it as well. We then would have the option to use "center", "side" layouts.

That's also a fantastic option to have, iirc there was a fork that had different layouts like that.
 <br>
PD: Just noticed that the `feaure request` label has a typo, it's missing the `t`.

**Comment by Samillion** [comment]
>  PD: Just noticed that the `feaure request` label has a typo, it's missing the `t`.

🤣 Oh my god, thank you for that. Fixed.

**Comment by Samillion** [comment]
Damn, making it thicker looks so much better: https://github.com/Samillion/ModernZ/issues/179#issuecomment-2460053701

**Comment by nekoxuee** [comment]
After reading #189 and noticing that a lot of users miss important details when posting issues, I think it could be helpful to create an issue template. What do you think?

**Comment by Samillion** [comment]
Fully agreed. The one thing I can't decide on is what to include in the issue template.

**Comment by Samillion** [comment]
Maybe adapt these for ModernZ? (from mpv repo)
- https://github.com/mpv-player/mpv/blob/master/.github/ISSUE_TEMPLATE/3_bug_report.yml
- https://github.com/mpv-player/mpv/blob/master/.github/ISSUE_TEMPLATE/5_feature_request.yml
- https://github.com/mpv-player/mpv/blob/master/.github/ISSUE_TEMPLATE/config.yml

Additionally, we can add two more for:
- Question(s)
- Discussion(s)

Then label each accordingly?

**Comment by nekoxuee** [comment]
Those look solid

**Comment by Samillion** [comment]
Issue templates are now active. 🎉

**Comment by Samillion** [comment]
@Xurdejl inspired by your discussion here, I've decided to also include
- https://github.com/Samillion/ModernZ/blob/main/.github/PULL_REQUEST_TEMPLATE
- https://github.com/Samillion/ModernZ/blob/main/.github/CONTRIBUTING.md
- https://github.com/Samillion/ModernZ/blob/main/.github/CODE_OF_CONDUCT.md

Also, adjusted [credits](https://github.com/Samillion/ModernZ#credits) to be more detailed

Which I think alongside the docs that already exist, covers everything.

**Comment by nekoxuee** [comment]
That sounds good.

Would it also be beneficial to include a style guide to ensure consistent code formatting and style conventions?

**Comment by Samillion** [comment]
Agreed. Though I think it's a bit early, as the code is still hectic from old forks. My next major PR (tomorrow or so), will be about simplifying logic, specifically in elements and layouts.

Also going to rename a lot of internal variables, labels. (nothing breaking though)

I think once we unify all or most of the code to one standard, we can safely add guidelines.

**Comment by sivateja09** [comment]
Hello! I don't know if this is the right place for stuff like this, but is it possible to add a fade in animation while showing osc similar to how it fades out?

I tried this and it worked

```
local function show_osc()
    if not state.enabled then return end

    msg.trace("show_osc")
    state.showtime = mp.get_time()

    if (user_opts.fadeduration > 0) then
        if state.osc_visible == false then
            state.anitype="in"
            osc_visible(true)
            request_tick()
        end
    end
end
```

P.S - Thank you for making an amazing osc!

**Reply by Samillion** [reply]
Will implement this today, thank you very much.

> Hello! I don't know if this is the right place for stuff like this

You can suggest things like this as a feature request in issues.

**Reply by Samillion** [reply]
Follow up here please: https://github.com/Samillion/ModernZ/issues/343
